File: src/pokeevaluator/data/natures.py
"""All 25 Gen III natures with stat modifiers."""

from __future__ import annotations

from dataclasses import dataclass

from pokeevaluator.data.gen3 import NATURE_BOOST, NATURE_NEUTRAL, NATURE_PENALTY
from pokeevaluator.models.pokemon import StatBlock, StatName


@dataclass(frozen=True, slots=True)
class Nature:
    name: str
    boost: StatName | None  # None for neutral natures
    penalty: StatName | None

    def modifiers(self) -> StatBlock:
        """Return a StatBlock of nature multipliers (HP is always 1.0)."""
        mods = {s: NATURE_NEUTRAL for s in StatName}
        if self.boost and self.penalty:
            mods[self.boost] = NATURE_BOOST
            mods[self.penalty] = NATURE_PENALTY
        return StatBlock(
            hp=mods[StatName.HP],
            atk=mods[StatName.ATK],
            def_=mods[StatName.DEF],
            spatk=mods[StatName.SPATK],
            spdef=mods[StatName.SPDEF],
            spe=mods[StatName.SPE],
        )


# fmt: off
NATURES: dict[str, Nature] = {n.name: n for n in [
    Nature("Hardy",   None,           None),
    Nature("Lonely",  StatName.ATK,   StatName.DEF),
    Nature("Brave",   StatName.ATK,   StatName.SPE),
    Nature("Adamant", StatName.ATK,   StatName.SPATK),
    Nature("Naughty", StatName.ATK,   StatName.SPDEF),
    Nature("Bold",    StatName.DEF,   StatName.ATK),
    Nature("Docile",  None,           None),
    Nature("Relaxed", StatName.DEF,   StatName.SPE),
    Nature("Impish",  StatName.DEF,   StatName.SPATK),
    Nature("Lax",     StatName.DEF,   StatName.SPDEF),
    Nature("Timid",   StatName.SPE,   StatName.ATK),
    Nature("Hasty",   StatName.SPE,   StatName.DEF),
    Nature("Serious", None,           None),
    Nature("Jolly",   StatName.SPE,   StatName.SPATK),
    Nature("Naive",   StatName.SPE,   StatName.SPDEF),
    Nature("Modest",  StatName.SPATK, StatName.ATK),
    Nature("Mild",    StatName.SPATK, StatName.DEF),
    Nature("Quiet",   StatName.SPATK, StatName.SPE),
    Nature("Bashful", None,           None),
    Nature("Rash",    StatName.SPATK, StatName.SPDEF),
    Nature("Calm",    StatName.SPDEF, StatName.ATK),
    Nature("Gentle",  StatName.SPDEF, StatName.DEF),
    Nature("Sassy",   StatName.SPDEF, StatName.SPE),
    Nature("Careful", StatName.SPDEF, StatName.SPATK),
    Nature("Quirky",  None,           None),
]}
# fmt: on


def get_nature(name: str) -> Nature:
    """Look up a nature by name (case-insensitive)."""
    key = name.capitalize()
    if key not in NATURES:
        raise ValueError(f"Natureza desconhecida: {name!r}")
    return NATURES[key]


def get_nature_modifiers(name: str) -> StatBlock:
    """Shortcut: return the modifier StatBlock for a nature name."""
    return get_nature(name).modifiers()

File: web/src/lib/data/natures.ts
import { Nature, StatBlock, StatName } from "../types";
import { NATURE_BOOST, NATURE_NEUTRAL, NATURE_PENALTY } from "./gen3";

function nature(name: string, boost: StatName | null, penalty: StatName | null): Nature {
  return { name, boost, penalty };
}

export const NATURES: Record<string, Nature> = {
  Hardy:   nature("Hardy",   null,          null),
  Lonely:  nature("Lonely",  StatName.ATK,  StatName.DEF),
  Brave:   nature("Brave",   StatName.ATK,  StatName.SPE),
  Adamant: nature("Adamant", StatName.ATK,  StatName.SPATK),
  Naughty: nature("Naughty", StatName.ATK,  StatName.SPDEF),
  Bold:    nature("Bold",    StatName.DEF,  StatName.ATK),
  Docile:  nature("Docile",  null,          null),
  Relaxed: nature("Relaxed", StatName.DEF,  StatName.SPE),
  Impish:  nature("Impish",  StatName.DEF,  StatName.SPATK),
  Lax:     nature("Lax",     StatName.DEF,  StatName.SPDEF),
  Timid:   nature("Timid",   StatName.SPE,  StatName.ATK),
  Hasty:   nature("Hasty",   StatName.SPE,  StatName.DEF),
  Serious: nature("Serious", null,          null),
  Jolly:   nature("Jolly",   StatName.SPE,  StatName.SPATK),
  Naive:   nature("Naive",   StatName.SPE,  StatName.SPDEF),
  Modest:  nature("Modest",  StatName.SPATK, StatName.ATK),
  Mild:    nature("Mild",    StatName.SPATK, StatName.DEF),
  Quiet:   nature("Quiet",   StatName.SPATK, StatName.SPE),
  Bashful: nature("Bashful", null,          null),
  Rash:    nature("Rash",    StatName.SPATK, StatName.SPDEF),
  Calm:    nature("Calm",    StatName.SPDEF, StatName.ATK),
  Gentle:  nature("Gentle",  StatName.SPDEF, StatName.DEF),
  Sassy:   nature("Sassy",   StatName.SPDEF, StatName.SPE),
  Careful: nature("Careful", StatName.SPDEF, StatName.SPATK),
  Quirky:  nature("Quirky",  null,          null),
};

export const NATURE_LIST = Object.values(NATURES);

export function getNature(name: string): Nature {
  const key = name.charAt(0).toUpperCase() + name.slice(1).toLowerCase();
  const n = NATURES[key];
  if (!n) throw new Error(`Unknown nature: ${name}`);
  return n;
}

export function getNatureModifiers(name: string): StatBlock {
  return natureModifiers(getNature(name));
}

export function natureModifiers(nature: Nature): StatBlock {
  const mods: Record<string, number> = {
    hp: NATURE_NEUTRAL,
    atk: NATURE_NEUTRAL,
    def_: NATURE_NEUTRAL,
    spatk: NATURE_NEUTRAL,
    spdef: NATURE_NEUTRAL,
    spe: NATURE_NEUTRAL,
  };
  if (nature.boost && nature.penalty) {
    mods[nature.boost] = NATURE_BOOST;
    mods[nature.penalty] = NATURE_PENALTY;
  }
  return {
    hp: mods.hp,
    atk: mods.atk,
    def_: mods.def_,
    spatk: mods.spatk,
    spdef: mods.spdef,
    spe: mods.spe,
  };
}
